Persistence Marshaller

ryanemerson opened the pull request:
https://issues.jboss.org/browse/ISPN-7409

ryanemerson commented:
The following failures are fixed by this [commit](https://github.com/infinispan/protostream/commit/3d13f20679c10a7e31c35b5736fde75884ffdbd3) in the forthcoming Protostream release:
```
org.infinispan.test.integration.as.jms.SearchNewEntityJmsMasterSlaveUsingInfinispanAs2ndCacheAndModulesIT.org.infinispan.test.integration.as.jms.SearchNewEntityJmsMasterSlaveUsingInfinispanAs2ndCacheAndModulesIT
org.infinispan.test.integration.as.jms.infinispan.SearchNewEntityJmsMasterSlaveUsingInfinispanAndModulesIT.org.infinispan.test.integration.as.jms.infinispan.SearchNewEntityJmsMasterSlaveUsingInfinispanAndModulesIT
org.infinispan.test.integration.as.wildfly.DirectoryProviderModuleMemberRegistrationIT.org.infinispan.test.integration.as.wildfly.DirectoryProviderModuleMemberRegistrationIT
org.infinispan.test.integration.as.wildfly.InfinispanModuleMemberRegistrationIT.org.infinispan.test.integration.as.wildfly.InfinispanModuleMemberRegistrationIT
```

ryanemerson commented:
@galderz Agreed that we need some kind of test/build-step to ensure that any changes are backwards compatible. There is already tooling out there todo this for static `*.proto` files, so I will investigate a solution for our generated proto files. Ideally it'll be something that is executed on a per-module basis, but worst case an integration test could be used for the entire build.

danberindei commented:
@ryanemerson needs rebase. Also, we should run PerfAck, even without stores.

ryanemerson commented:
> * I would have liked to have an explicit user marshaller, not just `PersistenceMarshaller.marshallUserObject/unmarshallUserBytes`

Do you mean something like `PersistenceUserMarshaller` that simply has the wrapper class and an instance of the configured user marshaller?



> * Ideally I would have liked each cache to have its own user marshaller, so that our key/value types don't mix with the user types unless they want to. Of course, atomic maps can exist in any cache so they break my ideal scenario.

If AtomicMaps didn't exist we wouldn't have to worry about that :stuck_out_tongue_winking_eye: 
But more seriously, I'm not sure what you mean by `our key/value types don't mix with the user types unless they want to`, how does having a user-marshaller per cache solve that issue? also, remember that user marshaller configuration is currently global not per cache.



> * Probably off-topic for your PR, but I was hoping this would be a chance to abstract away the persistent state from memory storage, so we could change memory storage and restart with the same store, or similarly add/remove indexing and restart with the same store (switching in memory between `ByteArrayWrapper` and `ProtobufValueWrapper`).

I think that's a good goal to aim for, but I think that's out of scope for this work considering how long this has been going on for.

ryanemerson commented on core/src/main/java/org/infinispan/marshall/persistence/impl/PersistenceContextInitializer.java:
```diff
@@ -0,0 +1,44 @@
+            EmbeddedMetadata.EmbeddedLifespanExpirableMetadata.class,
+            EmbeddedMetadata.EmbeddedMaxIdleExpirableMetadata.class,
+            EventLogCategory.class,
+            EventLogLevel.class,
```
Whilst they are values in the cache, they're internal types in a similar way to EntryVersion etc, so I think it makes sense if we control how they are marshalled. Furthermore, if the user configures there own marshaller, they would have to implement a means to marshall these classes if they ever wanted to make use of the event-log

danberindei replied:
Fair enough

anistor commented on core/src/main/java/org/infinispan/atomic/impl/AtomicKeySetImpl.java:
```diff
@@ -554,4 +557,88 @@ public Externalizable readObject(ObjectInput input) throws IOException, ClassNot
       }
    }
+
+   public static class KeyMarshaller implements MessageMarshaller<AtomicKeySetImpl.Key> {
```
Was wondering why we use a manually written marshaller for this? Any protostream bug that had to be worked around?

Please keep in mind that manual marshallers have the potential of being 10x slower than generated marshallers.  IMO, we should deprecate MessageMarshaller in next major protostream vesion.

ryanemerson replied:
Hmm I can't remember, but I'm hoping that https://github.com/infinispan/infinispan/pull/6969 will get integrated so that we can remove all of the manual marshallers anyway. 

A tentative +1 for removing manual marshallers in future protostream versions. An obvious objection would be if AtomicMaps don't get removed, but another concern is that they might be necessary in some obscure part of our code when converting the GlobalMarshaller to utilise protostream.

danberindei replied:
@anistor I'm guessing it's because it must use `persistenceMarshaller.marshallUserObject`.

Could we write custom marshallers that use the same approach as the generated marshallers for speed? I'm thinking we could keep fields immutable if we read all the fields from the message in a loop, saved them in local variables, and invoked a constructor at the end.

OTOH I would very much like to replace `MessageMarshaller` with something else, I don't like how overloaded `Marshaller` has become in our code base with this PR.

ryanemerson replied:
@danberindei In terms of this PR, now that AtomicMaps have been removed I don't think we need any manual marshaller implementations.

I do like the idea of being able to write performant marshallers by hand though.

anistor replied:
For now I would simply like to avoid manually written marshallers if possible. This is not the right place to discuss protostream extensions.

anistor commented on remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/ProtobufMetadataManagerInterceptor.java:
```diff
@@ -127,7 +126,7 @@ public Object visitPutKeyValueCommand(InvocationContext ctx, PutKeyValueCommand
                   .addProtoFile(key, (String) command.getValue());
             try {
                serializationContext.registerProtoFiles(source);
-            } catch (IOException | DescriptorParserException e) {
```
This change in protostream 4.3 (IOException no longer declared to be thrown) is regrettably not backward compatible with 4.2.x, so I will undo it in next 4.3.x alpha release. For now, keep this as is.

danberindei replied:
> 13.4.21. Method and Constructor Throws
>
> Changes to the throws clause of methods or constructors do not break compatibility with pre-existing binaries; these clauses are checked only at compile time.

anistor replied:
Was not worried about binary compat here. This change breaks compilation.

anistor commented:
Was this rebased to latest master @ryanemerson ?  Also, commit name "Upgrade to protostream 4.3.0.Alpha5" should say Alpha6

ryanemerson commented:
> Was this rebased to latest master @ryanemerson ?

Not since AtomicMap was removed. On it

anistor commented on core/src/main/java/org/infinispan/marshall/persistence/impl/PersistenceContextInitializer.java:
```diff
@@ -0,0 +1,43 @@
+            PersistenceMarshallerImpl.UserBytes.class,
+            SimpleClusteredVersion.class,
+      },
+      schemaFileName = PACKAGE_NAME + ".gen.proto",
```
.gen  was needed here in order to distinguish between hand written and generated schemas. This is no longer needed after the removal of the atomic objects @ryanemerson

danberindei commented:
> > * I would have liked to have an explicit user marshaller, not just `PersistenceMarshaller.marshallUserObject/unmarshallUserBytes`
> 
> Do you mean something like `PersistenceUserMarshaller` that simply has the wrapper class and an instance of the configured user marshaller?
> 

Looking at the code again, I think what I wanted was basically to rename `PersistenceMarshaller` to `UserMarshaller` :)

I was hoping to put hard barrier between our types and user types, so the decision whether to switch to another marshaller would be done only in specific places in our MessageMarshallers and not every time writeObject/readObject is called. But I realize that may be messy, if there are a lot more places where the switch is needed than I'm imagining now.

> > * Ideally I would have liked each cache to have its own user marshaller, so that our key/value types don't mix with the user types unless they want to. Of course, atomic maps can exist in any cache so they break my ideal scenario.
> 
> If AtomicMaps didn't exist we wouldn't have to worry about that stuck_out_tongue_winking_eye

Actually it's almost the same with entry metadata, I believe we can have either fully custom metadata, fully standard metadata, or standard metadata with custom version.

> But more seriously, I'm not sure what you mean by `our key/value types don't mix with the user types unless they want to`, how does having a user-marshaller per cache solve that issue? also, remember that user marshaller configuration is currently global not per cache.
> 

I was imagining each cache's user marshaller to start with an empty serialization context and include only the classes the user adds.
It would kind of require the ability to set a custom marshaller per cache as well, and I was also ignoring the metadata that goes in the same cache.

> > * Probably off-topic for your PR, but I was hoping this would be a chance to abstract away the persistent state from memory storage, so we could change memory storage and restart with the same store, or similarly add/remove indexing and restart with the same store (switching in memory between `ByteArrayWrapper` and `ProtobufValueWrapper`).
> 
> I think that's a good goal to aim for, but I think that's out of scope for this work considering how long this has been going on for.

Agree.

danberindei commented on persistence/remote/src/test/java/org/infinispan/persistence/remote/RemoteStoreConfigTest.java:
```diff
@@ -35,6 +36,7 @@
    public void startUp() {
       cacheManager = TestCacheManagerFactory.createCacheManager(hotRodCacheConfiguration());
       assertEquals(cacheManager.getCache().size(), 0);
+      TestingUtil.registerTranscoders(cacheManager, new ProtostreamBinaryTranscoder());
```
I'm surprised the server doesn't already do this.

ryanemerson replied:
It's normally added by `remote-query-server`'s LifecycleManager, however that's not called when creating the HotrodServer via `HotRodTestingUtil.startHotRodServer`

danberindei replied:
I'm still surprised... `HotRodTestingUtil.startHotRodServer` doesn't do anything to skip the initialization of certain modules.

ryanemerson replied:
@danberindei Neither the remote store module, or  `infinispan-server-hotrod` has a dependency on `remote-query-server`, therefore it's not possible for the `RemoteStoreConfigTest` to have the query server's LifecycyleManager executed.

danberindei replied:
`persistence-remote` could add `remote-query-server` as a test dependency. 

But query isn't really needed to support the protobuf media type on the server, so `ProtostreamBinaryTranscoder` and the other transcoders in the same package could and IMO should be registered by `org.infinispan.server.core.LifecycleCallbacks`.

ryanemerson replied:
I've moved the `ProtostreamBinaryTranscoder` to server-core, however it does not make sense for the other transcoders to be moved as they rely on the `ProtobufMetadataManager`'s SerializationContext.

danberindei commented on server/integration/event-logger/src/main/java/org/infinispan/server/eventlogger/ServerEventImpl.java:
```diff
@@ -20,17 +14,24 @@
+   Optional<String> detail;
+   Optional<String> context;
+   Optional<String> who;
+   Optional<String> scope;
```
IntelliJ warns about Optional fields anyway, so maybe it's better to keep these fields regular strings, move the annotations here, and make the methods return `Optional.ofNullable()`

danberindei commented on server/integration/event-logger/src/main/java/org/infinispan/server/eventlogger/ServerEventImpl.java:
```diff
@@ -20,17 +14,24 @@
-   private final Optional<String> context;
-   private final Optional<String> who;
-   private final Optional<String> scope;
+   @ProtoField(number = 1, required = true)
```
For backwards compatibility we shouldn't make anything required, IIRC the next version of Protobuf won't have required fields at all.

danberindei commented on tasks/scripting/src/main/java/org/infinispan/scripting/impl/ScriptMetadata.java:
```diff
@@ -27,33 +25,58 @@
+   MediaType dataType;
+
+   private Optional<String> language;
+   private Optional<String> role;
```
Same here, we can make only the methods deal with `Optional`

danberindei commented on tools/src/main/java/org/infinispan/tools/store/migrator/StoreMigrator.java:
```diff
@@ -20,25 +44,64 @@
    private static final int DEFAULT_BATCH_SIZE = 1;
+   // A list of all internal classes that were previously marshallable, but we no longer provide
+   // guarantees over byte compatibility in future versions
+   private static final Set<Class<?>> INTERNAL_BLACKLIST = new HashSet<>();
```
I'm not sure I would have read the store migrator changes anyway, but it would have been nice it they were in a separate PR, because GitHub collapses all of them.

ryanemerson replied:
Agreed, it does introduce a lot of noise. I'll issue the commit as a separate PR once this is merged.

ryanemerson commented on core/src/main/java/org/infinispan/marshall/persistence/PersistenceMarshaller.java:
```diff
@@ -0,0 +1,39 @@
+    * @param initializer whose schemas and marshallers' will be registered with the {@link PersistenceMarshaller} {@link
+    *                    SerializationContext}
+    */
+   void init(SerializationContextInitializer initializer);
```
how about `register()`? Because we're registering the `SerializationContextInitializer`'s schemas/marshallers to the PMs SerializationContext?

ryanemerson commented on core/pom.xml:
```diff
@@ -179,6 +185,14 @@
+            <includes>
+               <include>generated/*.proto</include>
+            </includes>
+            <targetPath>schema</targetPath>
```
I tried todo this, however I run into issues with checkstyle now checking the generated `Log.i18n.properties` file and failing. I tried reading the entire build dir and excluding paths with no luck. So I figured narrowing the search to just `schema/` dir and using `targetPath` was acceptable.

danberindei replied:
@ryanemerson I didn't know we had checkstyle rules for properties files, but why would checkstyle ever check files in `target/classes`? For ISPN-10137 I'm generating source files with plenty of unused imports in `target/generated-test-sources`, and checkstyle never complained.

danberindei replied:
To answer my own question, it's probably because I never added `target/generated-test-sources` as an explicit source directory in Maven, the compiler plugin added it, and that's why Checkstyle didn't include it. So perhaps the solution in your case is to remove the `<resource>` tag, because everything in `target/classes` is included in the jar anyway.

Oh, and perhaps we should move the Checkstyle execution before compilation.

ryanemerson replied:
@danberindei For some reason with core the generated proto files are not included in the jar without my `<resource>` definition :confused:

danberindei replied:
Sorry, I have a really short memory: I had to add the `META-INF/services` directory to the `<Include-Resource>` bundle instructions configuration myself when working on ISPN-10137 :) It's probably because the regular jar plugin doesn't care what's in `target/classes`, but `maven-bundler-plugin` is more anal about including only .class files and the resources that `maven-resource-plugin` saw in the `resource` phase (which is before `compile`)

Since we already use the `<Include-Resource>` hack for other resources, I'd use it for proto schemas as well, and it should take care of checkstyle.

ryanemerson replied:
I've also applied this to `infinispan-lucene-directory`, `infinispan-remote-query-server` and `infinispan-server-memcached`.

ryanemerson commented on core/pom.xml:
```diff
@@ -165,6 +170,7 @@
             <filtering>false</filtering>
             <includes>
                <include>schema/*-${infinispan.core.schema.version}.xsd</include>
+               <include>schema/*.proto</include>
```
Updated to have:

```
schema/
schema/infinispan-config-10.0.xsd
schema/proto/
schema/proto/generated/
schema/proto/generated/persistence.core.proto
```

danberindei replied:
Why not `proto/persistence.core.proto`, without the `schema` and `generated` directories?

ryanemerson replied:
I included it in `schema`, because it's a schema. I can add it as just `root` if you prefer, but I don't think it makes any real difference.

The generated dir is included to allow us to be able to easily discern between generated and manually specified schemas. This was more of an issue when we had the AtomicMap marshallers etc, but I figured it was best to adopt an approach from the start that allows this kind of flexibility. I suppose we could remove the dir and just utilise the .proto filename.

danberindei replied:
Ok, let's put them in `schema`, although https://developers.google.com/protocol-buffers/docs/proto never says "schema" ;)

But I'm really not convinced about the potential for confusion with non-generated proto files, since the generated files are all in `target`.

ryanemerson replied:
> But I'm really not convinced about the potential for confusion with non-generated proto files, since the generated files are all in `target`.

I was thinking more in terms of the jar's content

ryanemerson replied:
So in the end I went with `proto/` like you suggested :smile:

ryanemerson commented on core/pom.xml:
```diff
@@ -210,6 +224,8 @@
                      !${project.groupId}.multimap.*,
                      !${project.groupId}.lock.*,
+                     !${project.groupId}.protostream.*,
+                     !${project.groupId}.*.generated.*,
```
Good spot, this is a hangover from before

ryanemerson commented on commons/src/main/java/org/infinispan/commons/marshall/PersistenceContextInitializer.java:
```diff
@@ -0,0 +1,25 @@
+      includeClasses = {
+            ByteBufferImpl.class,
+            MediaType.class,
+            WrappedByteArray.class
```
`WrappedByteArray` needs to be written to the store as we need to be able to distinguish between arbitrary user bytes that have been marshalled (UserObject) and bytes which should simply be returned as stored.

ryanemerson replied:
I've updated MarshalledValueImpl to write `bytes` directly, so `ByteBufferImpl` serialization is no longer necessary.

danberindei replied:
> WrappedByteArray needs to be written to the store as we need to be able to distinguish between arbitrary user bytes that have been marshalled (UserObject) and bytes which should simply be returned as stored.

I was asking because or my pet peeve with the store format changing because of a cache configuration change, as `WrappedByteArray` isn't always the result of the application inserting a `byte[]` in the cache, it's also sometimes because `BinaryEncoder` or `MarshallerEncoder` replaced the application-supplied value with a `byte[]`. I guess that's an encoding problem though, not a marshalling problem.

ryanemerson commented on core/src/main/java/org/infinispan/persistence/spi/MarshalledValue.java:
```diff
@@ -10,11 +12,16 @@
  */
 public interface MarshalledValue {
+
+   @ProtoField(number = 1, name = "value", javaType = ByteBufferImpl.class)
```
Yes. Originally I was thinking that by adding it to the SPI it would mean that the proto format would be applicable to custom implementations of the interface, but we have to register the concrete class with the PersistenceContextInitializer anyway, so this doesn't work.

ryanemerson commented on commons/src/main/java/org/infinispan/commons/marshall/StreamAwareMarshaller.java:
```diff
@@ -0,0 +1,61 @@
+
+/**
+ * An interface that facilitates the marshalling/unmarshalling of objects from
+ * the provided {@link java.io.OutputStream}/{@link java.io.InputStream}.
```
> And since the package is `@public`, do we expect users to use or implement it?

I think this depends on whether we want users to be able to provide their own marshallers or not as we discussed on Zulip. If we're going to continue to support this, then I think we should make this interface public and make it the preferred interface for users to implement as it's the most simple. However, if we're going to deprecate custom user marshallers then I don't see any reason for it to be public as the `InitializationContext` exposes the `PersistenceMarshaller` interface which extends `StreamAwareMarshaller`.

Internally, we should be aiming to use `StreamAwareMarshaller` and long term we should try to avoid calls to `objectToByteBuffer` etc where possible.

danberindei replied:
@ryanemerson because of the `@public` tag in package-info.java, all the classes in `org.infinispan.commons.marshall` are already public, so you should add an explicit `@private` tag if you don't want it to be public (yet). Alternatively you could remove the `@public` tag from the package and add it only to the classes which really need to be public.

ryanemerson commented on core/src/main/java/org/infinispan/marshall/core/JBossUserMarshaller.java:
```diff
@@ -0,0 +1,63 @@
+
+   /**
+    * A {@link org.jboss.marshalling.ObjectTable} implementation that creates {@link org.jboss.marshalling.ObjectTable.Writer}
+    * based upon a users configured {@link org.infinispan.commons.marshall.Externalizer} implementations.
```
I think we should wait until the config is in place for a user to configure Protostream annotated classes etc. My plan was to add that in a subsequent PR with related tests.

ryanemerson replied:
https://issues.jboss.org/browse/ISPN-10280

ryanemerson commented on core/src/main/java/org/infinispan/marshall/persistence/impl/PersistenceMarshallerImpl.java:
```diff
@@ -0,0 +1,246 @@
+
+      try {
+         boolean wrapped = isPersistenceClass(o);
+         o = wrap(o, wrapped);
```
I agree, I've changed it to the following:
```java
boolean requiresWrapping = !isPersistenceClass(o);
if (requiresWrapping)
    o = wrapUserObject(o);
```

ryanemerson commented on core/src/test/java/org/infinispan/partitionhandling/NumOwnersNodeCrashInSequenceTest.java:
```diff
@@ -50,6 +53,9 @@ protected void createCacheManagers() throws Throwable {
       configBuilder.clustering().hash().numSegments(4).stateTransfer().timeout(30000);
+      globalBuilder = GlobalConfigurationBuilder.defaultClusteredBuilder();
+      // Must explicitly add with id >= 2500 so that the user marshaller is able to store it
+      globalBuilder.serialization().addAdvancedExternalizer(2500, new JGroupsAddress.Externalizer());
```
Nope, the ID specified in the config always takes precedent over the value returned by `getId`

danberindei replied:
Is it because the id from `Externalizer.getId()` is added to the internal externalizers map and the user marshaller only sees the externalizers map? And why >= 2500?

ryanemerson replied:
`> 2500` because we have already reserved up until `2149` according to the docs and in fact we have actually gone higher when you look at JCache Ids.

https://infinispan.org/docs/stable/user_guide/user_guide.html#preassigned_externalizer_id_ranges

ryanemerson replied:
The precedent of utilising the configured id vs `getId` is already established, I haven't changed anything: https://github.com/infinispan/infinispan/blob/4d1800b1232c17aa63b6a34acdc34865ff2af5c8/core/src/main/java/org/infinispan/marshall/core/ExternalExternalizers.java#L36-L38

ryanemerson commented on core/src/test/java/org/infinispan/persistence/support/AsyncStoreTest.java:
```diff
@@ -108,9 +108,9 @@ private void createStore(boolean slow) throws PersistenceException {
-      writer.start();
       loader = new AdvancedAsyncCacheLoader(underlying, writer.getState());
       loader.init(ctx);
+      writer.start();
```
Not anymore, this is a hangover from when marshallers were being generated at runtime.

ryanemerson commented on core/src/main/java/org/infinispan/util/ByteString.java:
```diff
@@ -52,22 +54,31 @@ public boolean equals(Object o) {
+      return bytes;
+   }
+
+   public void setBytes(byte[] bytes) {
```
I have made `setBytes` be package-private as that works with the latest protostream changes. We can remove this in the future when protostream supports immutable objects.

ryanemerson commented on lucene/lucene-directory/src/main/java/org/infinispan/lucene/impl/FileListCacheValue.java:
```diff
@@ -40,6 +35,29 @@ public FileListCacheValue() {
+      try {
+         return filenames;
+      } finally {
+         readLock.unlock();
```
`getFileNames` was introduced just for protostream, it isn't called anywhere else, so I think the readlock logic is the same as when used in the old Externalizer impl.

danberindei replied:
The externalizer was holding a read lock while writing the filenames, not just while reading the field:

https://github.com/ryanemerson/infinispan/blob/ee9f8bab2079a9cc44cb5ea0c458b15f47a7a7d4/lucene/lucene-directory/src/main/java/org/infinispan/lucene/impl/FileListCacheValue.java#L168

ryanemerson replied:
@danberindei I see what you mean. So we could defensively copy `filenames` whilst holding the read lock and the return the copy via `getFileNames`.

ryanemerson commented on core/src/test/java/org/infinispan/marshall/InvalidatedMarshalledValueTest.java:
```diff
@@ -30,6 +31,9 @@ protected void createCacheManagers() throws Throwable {
       createClusteredCaches(2, "invlSync", invlSync);
+      TestingUtil.initJbossMarshallerTypeHints(manager(0), new InvalidatedPojo());
+      TestingUtil.initJbossMarshallerTypeHints(manager(1), new InvalidatedPojo());
+      InvalidatedPojo.invalidSerializationCount = 0;
```
See https://github.com/ryanemerson/infinispan/blob/2b58072b07f4fbb1e6d72b25464f73feb6e74b86/core/src/test/java/org/infinispan/test/TestingUtil.java#L1953-L1956

danberindei replied:
@ryanemerson I saw that comment, but it doesn't mention the persistence marshaller at all, and what I didn't understand was why there are more `isMarshallable()` checks with the persistence marshaller than without it.

ryanemerson replied:
@danberindei It definitely does mention the PersistenceMarshaller :smile:

But the additional `isMarshaller` check is introduced [here](https://github.com/ryanemerson/infinispan/blob/4d1800b1232c17aa63b6a34acdc34865ff2af5c8/core/src/main/java/org/infinispan/marshall/core/JbossInternalObjectTable.java#L48). It's necessary to see whether the GlobalMarshaller can delegate to the PersistenceMarshaller. If not, then null is returned and jboss-marshalling is used.

ryanemerson commented:
@danberindei Feedback addressed in https://github.com/infinispan/infinispan/pull/6906/commits/50f5c37b2de306f4fc5bb28f23bc25491fbcc0f1

I have removed the store migrator changes from this PR and will issue a separate one, work tracked here: https://issues.jboss.org/browse/ISPN-10276

Similarly, I have also removed the `proto-schema-compatibility-plugin` from this PR and will issue later. Work tracked here: https://issues.jboss.org/browse/ISPN-10246

ryanemerson commented:
run performance tests please

ryanemerson commented:
run performance tests please

ryanemerson commented:
@danberindei @anistor No noticeable regression in performance.

ryanemerson commented:
I think I have addressed all concerns. If you're happy with the changes, let me know and I'll squash the commits.

danberindei commented on commons/pom.xml:
```diff
@@ -14,6 +14,17 @@
+      <dependency>
+         <groupId>org.infinispan.protostream</groupId>
+         <artifactId>protostream-processor</artifactId>
+         <scope>compile</scope>
```
```suggestion
         <scope>provided</scope>
```

ryanemerson replied:
@danberindei This results in non-commons modules not compiling.

danberindei replied:
Yes, you need to add it to all the modules that use the annotation processor, or you could add it to the parent.

But with `compile` all the applications using `commons` will end up with the `protostream-processor` on their runtime classpath, that's why I moved `component-processor` to the `provided` scope as well:

https://github.com/infinispan/infinispan/commit/8994b1ee6137a6c1a6dde0fd2adc9cf6e50d16d8
https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html#Dependency_Scope

ryanemerson commented:
I've reverted back to protostream `Alpha6` as 7 introduced test failures

danberindei commented on core/src/main/java/org/infinispan/marshall/persistence/impl/PersistenceMarshallerImpl.java:
```diff
@@ -0,0 +1,234 @@
+      if (requiresWrapping)
+         o = wrapUserObject(o);
+      int size = sizeEstimate(o, true);
+      ProtobufUtil.toWrappedStream(serializationContext, out, o, size);
```
I really like the way protostream has the stream parameter before the object, not sure how we ended up with the wrong order in our interfaces ;)

ryanemerson commented on core/src/main/java/org/infinispan/marshall/persistence/PersistenceMarshaller.java:
```diff
@@ -0,0 +1,24 @@
+ * @author Ryan Emerson
+ * @since 10.0
+ */
+public interface PersistenceMarshaller extends Marshaller, StreamAwareMarshaller {
```
Unfortunately not, as most of our stores are dependent on the `Marshaller` interface via the `InitializationContext`.

danberindei replied:
Maybe you could extend `Marshaller` only in `PersistenceMarshallerImpl`?

ryanemerson replied:
That would mean that all of our stores would have to cast to `(PersistenceMarshallerImpl)` which I don't think is particularly appealing.

ryanemerson commented:
Squashed and commit messages updated.

danberindei commented on pom.xml:
```diff
@@ -2491,6 +2491,13 @@
             <type>test-jar</type>
             <scope>test</scope>
          </dependency>
+          <dependency>
```
This is exactly why I stuck to the 3-space indent in all the POMs where I copied the `component-processor` dependency ;)

ryanemerson commented on core/src/main/java/org/infinispan/interceptors/impl/CacheWriterInterceptor.java:
```diff
@@ -499,6 +499,9 @@ public int getNumberOfPersistedEntries() {
    CompletionStage<Void> storeEntry(InvocationContext ctx, Object key, FlagAffectedCommand command, boolean incrementStats) {
+      if (persistenceManager.isReadOnly())
+         return CompletableFutures.completedNull();
+
```
@wburns FYI, this optimisation to avoid unnecessary marshalling was accidentally removed in ISPN-9722. It resulted in the `HotRodConfigurationTest` encountered on CI, as it meant that when the CacheLoader was active, the JGroupAddress was being marshalled even though there was no store, resulting in a `NotSerializableException`.

wburns replied:
Yeah, not sure how I removed that :( Sorry about that.

danberindei commented:
Integrated, thanks Ryan!
